File: frontend/src/api/world.ts
import { apiFetch } from "./client";
import type {
  InventorySummary,
  LocationGenerationPreview,
  Monster,
  NpcGenerationPreview,
  Scene,
  SceneCombatResult,
  SceneParticipant,
  WorldItem,
} from "./types";

type Items<T> = { items: T[] };

export type SiteGenerationInput = {
  site_type: "building" | "dungeon";
  name: string;
  brief: string;
  region_path: string;
  maximum_levels: number;
  rooms_min: number;
  rooms_max: number;
  party_level: number;
  party_size: number;
  character_ids: string[];
  starting_difficulty: "low" | "moderate" | "high";
  difficulty_growth: number;
  monster_density: number;
  reward_rate: number;
  overall_scale: "small" | "medium" | "large" | "huge";
  minimum_room_size: "small" | "medium" | "large" | "huge";
  maximum_room_size: "small" | "medium" | "large" | "huge";
  generate_npcs: boolean;
  generate_monsters: boolean;
  generate_loot: boolean;
  seed?: number;
};
export type SiteCell = { row: number; col: number; kind: string; label: string; blocks_sight?: boolean };
export type SiteRoomPlan = {
  room_index: number;
  name: string;
  room_type: string;
  description?: string | null;
  bounds?: Record<string, number>;
  bounds_json?: Record<string, number>;
  encounter_json?: {
    visibility?: "hidden" | "revealed";
    monsters?: Array<Record<string, unknown>>;
    npcs?: Array<Record<string, unknown>>;
  };
};
export type SiteLevelPreview = {
  level_index: number;
  name: string;
  description: string;
  difficulty: "low" | "moderate" | "high";
  encounter_budget_xp: number;
  reward_budget_gp: number;
  layout: { width: number; height: number; cell_size_ft: number; cells: SiteCell[] };
  visual_theme?: {
    theme?: string;
    label?: string;
    palette?: string;
    source_kind?: "preset" | "compiled";
    keywords?: string[];
    atmosphere?: string;
    hazard_motifs?: string[];
  };
  rooms: SiteRoomPlan[];
  connectors: Array<Record<string, unknown>>;
  monster_plan: Array<{ name: string; quantity: number; xp_each: number; source: string; room_index?: number }>;
  npc_plan?: Array<{ name: string; role: string; room_index: number }>;
  reward_plan: Array<{
    name: string;
    value_gp: number;
    category?: string;
    room_index?: number;
    source_kind?: string;
    rarity?: string | null;
  }>;
  quality?: {
    score: number;
    room_size_cv: number;
    largest_smallest_ratio: number;
    valid_connectors: number;
    algorithm: string;
  };
};
export type SiteGenerationPreview = {
  schema_version: string;
  site: SiteGenerationInput & { theme: string; seed: number; generation_parameters: Record<string, unknown> };
  region: { path: string[]; name: string };
  levels: SiteLevelPreview[];
  warnings: string[];
};
export type AdventureSite = SiteGenerationPreview["site"] & {
  id: string;
  version: number;
  region_map_id: string;
  location_id: string;
  map_position: { row: number; col: number };
  status: string;
  levels?: Array<SiteLevelPreview & { id: string }>;
};
export type RegionMap = {
  id: string; name: string; width: number; height: number;
  map_json: { pois?: Array<{ site_id: string; name: string; site_type: string; row: number; col: number }> };
};

export function previewSiteGeneration(
  campaignId: string,
  input: SiteGenerationInput,
): Promise<SiteGenerationPreview> {
  return apiFetch(`/campaigns/${campaignId}/sites/generate/preview`, {
    method: "POST",
    body: input,
  });
}

export function confirmSiteGeneration(
  campaignId: string,
  preview: SiteGenerationPreview,
  requestId: string,
): Promise<AdventureSite> {
  return apiFetch(`/campaigns/${campaignId}/sites/generate/confirm`, {
    method: "POST",
    body: { preview },
    headers: { "X-Request-ID": requestId },
  });
}

export async function listAdventureSites(campaignId: string, signal?: AbortSignal): Promise<AdventureSite[]> {
  return (await apiFetch<{ sites: AdventureSite[] }>(`/campaigns/${campaignId}/sites`, { signal })).sites;
}

export function getAdventureSite(campaignId: string, siteId: string, signal?: AbortSignal): Promise<AdventureSite> {
  return apiFetch(`/campaigns/${campaignId}/sites/${siteId}`, { signal });
}

export function setSiteRoomVisibility(
  campaignId: string,
  siteId: string,
  levelIndex: number,
  roomIndex: number,
  visible: boolean,
): Promise<{
  site_id: string;
  level_index: number;
  room_index: number;
  visibility: "hidden" | "revealed";
  tokens_updated: number;
}> {
  return apiFetch(
    `/campaigns/${campaignId}/sites/${siteId}/levels/${levelIndex}/rooms/${roomIndex}/visibility`,
    { method: "PUT", body: { visible } },
  );
}

export function deleteAdventureSite(
  campaignId: string,
  siteId: string,
  version: number,
): Promise<void> {
  return apiFetch(`/campaigns/${campaignId}/sites/${siteId}?version=${version}`, {
    method: "DELETE",
  });
}

export async function listRegionMaps(campaignId: string, signal?: AbortSignal): Promise<RegionMap[]> {
  return (await apiFetch<{ region_maps: RegionMap[] }>(`/campaigns/${campaignId}/region-maps`, { signal })).region_maps;
}

export function generateNpc(
  campaignId: string,
  input: {
    mode: "quick" | "guided";
    brief: string;
    answers: Record<string, string>;
  },
): Promise<NpcGenerationPreview> {
  return apiFetch(`/campaigns/${campaignId}/generate/npc`, {
    method: "POST",
    body: input,
  });
}

export function generateLocation(
  campaignId: string,
  input: { brief: string; maximum_depth: number; scale: "small" | "medium" | "large" },
): Promise<LocationGenerationPreview> {
  return apiFetch(`/campaigns/${campaignId}/generate/location`, {
    method: "POST",
    body: input,
  });
}

export function confirmLocation(
  campaignId: string,
  preview: LocationGenerationPreview,
): Promise<{ locations: unknown[]; items: WorldItem[] }> {
  return apiFetch(`/campaigns/${campaignId}/generate/location/confirm`, {
    method: "POST",
    body: { preview },
  });
}

export async function listWorldItems(
  campaignId: string,
  filters: { location_id?: string; owner_character_id?: string } = {},
  signal?: AbortSignal,
): Promise<WorldItem[]> {
  const query = new URLSearchParams(filters);
  return (
    await apiFetch<Items<WorldItem>>(
      `/campaigns/${campaignId}/items${query.size ? `?${query}` : ""}`,
      { signal },
    )
  ).items;
}

export function createWorldItem(
  campaignId: string,
  input: Partial<WorldItem> & Pick<WorldItem, "name">,
): Promise<WorldItem> {
  return apiFetch(`/campaigns/${campaignId}/items`, { method: "POST", body: input });
}

export function pickupItem(
  campaignId: string,
  itemId: string,
  input: { character_id: string; quantity: number; version: number },
): Promise<{ item: WorldItem; inventory: InventorySummary }> {
  return apiFetch(`/campaigns/${campaignId}/items/${itemId}/pickup`, {
    method: "POST",
    body: input,
  });
}

export function getInventory(
  campaignId: string,
  characterId: string,
  signal?: AbortSignal,
): Promise<InventorySummary> {
  return apiFetch(`/campaigns/${campaignId}/characters/${characterId}/inventory`, { signal });
}

export async function listMonsters(
  campaignId: string,
  signal?: AbortSignal,
): Promise<Monster[]> {
  return (await apiFetch<Items<Monster>>(`/campaigns/${campaignId}/monsters`, { signal })).items;
}

export function createMonster(
  campaignId: string,
  input: Partial<Monster> & Pick<Monster, "name">,
): Promise<Monster> {
  return apiFetch(`/campaigns/${campaignId}/monsters`, { method: "POST", body: input });
}

export async function listScenes(
  campaignId: string,
  signal?: AbortSignal,
): Promise<Scene[]> {
  return (await apiFetch<Items<Scene>>(`/campaigns/${campaignId}/scenes`, { signal })).items;
}

export function createScene(
  campaignId: string,
  input: { name: string; location_id?: string | null; description?: string | null; notes?: string | null },
): Promise<Scene> {
  return apiFetch(`/campaigns/${campaignId}/scenes`, { method: "POST", body: input });
}

export type PersistentSceneGrid = { id: string; width: number; height: number; cell_size_ft: number; mode: "narrative" | "exploration" | "combat"; public_description: string | null; dm_description: string | null; layers_json: Record<string, unknown> };
export type SceneToken = { id: string; label: string; row: number; col: number; entity_type: string; entity_id?: string | null; visible: boolean };
export type PersistentSceneObject = { id: string; version: number; label: string; row: number; col: number; object_type: string; state: string; visibility: string };
export function getSceneGrid(campaignId: string, sceneId: string, signal?: AbortSignal): Promise<{ grid: PersistentSceneGrid; tokens: SceneToken[]; objects: PersistentSceneObject[] }> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/grid`, { signal });
}
export function createPersistentGrid(campaignId: string, sceneId: string, input: Partial<PersistentSceneGrid> = {}): Promise<PersistentSceneGrid> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/grid`, { method: "POST", body: input });
}
export function createSceneObject(campaignId: string, sceneId: string, input: { object_type: "wall" | "door" | "cover" | "terrain" | "light" | "trap" | "treasure" | "furniture" | "portal"; label: string; row: number; col: number; visibility?: "public" | "dm" | "hidden"; interaction_json?: Record<string, unknown>; metadata_json?: Record<string, unknown> }): Promise<PersistentSceneObject> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/objects`, { method: "POST", body: input });
}
export type SceneTokenInput = { entity_type: "character" | "npc" | "monster" | "marker"; entity_id?: string | null; label: string; row: number; col: number; size_cells?: number; elevation_ft?: number; visible?: boolean; metadata_json?: Record<string, unknown> };
export function createSceneToken(campaignId: string, sceneId: string, input: SceneTokenInput): Promise<SceneToken> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/tokens`, { method: "POST", body: input });
}
export type ExplorationInput = { action: "move" | "search" | "interact" | "explore"; minutes: number; token_id?: string | null; path?: Array<[number, number]>; object_id?: string | null; object_state?: string | null; notes?: string | null; preview_token?: string; idempotency_key?: string };
export function previewExploration(campaignId: string, sceneId: string, input: ExplorationInput): Promise<Record<string, unknown>> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/exploration/preview`, { method: "POST", body: input });
}
export function confirmExploration(campaignId: string, sceneId: string, input: ExplorationInput): Promise<Record<string, unknown>> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/exploration/confirm`, { method: "POST", body: input });
}
export type TravelInput = { to_location_id: string; distance_miles: number; pace: "fast" | "normal" | "slow"; encounter?: { title: string; outcome: "avoided" | "resolved" | "evaded"; summary: string; visibility: "dm" | "players" }; notes?: string | null; preview_token?: string; idempotency_key?: string };
export function previewTravel(campaignId: string, input: TravelInput): Promise<Record<string, unknown>> {
  return apiFetch(`/campaigns/${campaignId}/travel/preview`, { method: "POST", body: input });
}
export function confirmTravel(campaignId: string, input: TravelInput): Promise<Record<string, unknown>> {
  return apiFetch(`/campaigns/${campaignId}/travel/confirm`, { method: "POST", body: input });
}

export type ResolutionVisibility = "dm" | "players";
export type ExplorationCharacterEffect = {
  character_id: string;
  character_version: number;
  damage?: number;
  max_hp_reduction?: number;
  condition_name?: string | null;
  condition_duration?: string | null;
  condition_notes?: string | null;
};
type ConfirmationFields = { preview_token: string; idempotency_key: string };
type Confirmed<T> = T & ConfirmationFields;

export type ChaseInput = {
  chase_event_id?: string | null;
  chase_version?: number | null;
  title: string;
  outcome: "success" | "failure";
  target_successes?: number;
  target_failures?: number;
  minutes?: number;
  summary: string;
  visibility?: ResolutionVisibility;
  character_effects?: ExplorationCharacterEffect[];
};
export const listChases = (campaignId: string, signal?: AbortSignal) =>
  apiFetch<Items<Record<string, unknown>>>(`/campaigns/${campaignId}/chases`, { signal }).then((result) => result.items);
export const previewChase = (campaignId: string, input: ChaseInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/chases/preview`, { method: "POST", body: input });
export const confirmChase = (campaignId: string, input: Confirmed<ChaseInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/chases/confirm`, { method: "POST", body: input });

export type TrapResolutionInput = {
  trap_version: number;
  outcome: "triggered" | "disarmed" | "bypassed" | "failed";
  result_state?: "active" | "disarmed" | "destroyed";
  minutes?: number;
  summary: string;
  visibility?: ResolutionVisibility;
  character_effects?: ExplorationCharacterEffect[];
};
export const previewTrapResolution = (campaignId: string, sceneId: string, trapId: string, input: TrapResolutionInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/scenes/${sceneId}/traps/${trapId}/preview`, { method: "POST", body: input });
export const confirmTrapResolution = (campaignId: string, sceneId: string, trapId: string, input: Confirmed<TrapResolutionInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/scenes/${sceneId}/traps/${trapId}/confirm`, { method: "POST", body: input });

export type EnvironmentHazardInput = {
  name: string;
  object_id?: string | null;
  object_version?: number | null;
  object_state?: "active" | "open" | "closed" | "destroyed" | "disarmed" | "picked_up" | null;
  minutes?: number;
  summary: string;
  visibility?: ResolutionVisibility;
  character_effects?: ExplorationCharacterEffect[];
};
export const previewEnvironmentHazard = (campaignId: string, sceneId: string, input: EnvironmentHazardInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/scenes/${sceneId}/hazards/preview`, { method: "POST", body: input });
export const confirmEnvironmentHazard = (campaignId: string, sceneId: string, input: Confirmed<EnvironmentHazardInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/scenes/${sceneId}/hazards/confirm`, { method: "POST", body: input });

export type AfflictionInput = {
  operation: "apply" | "progress" | "cure";
  character_id: string;
  character_version: number;
  condition_id?: string | null;
  condition_version?: number | null;
  affliction_type: "poison" | "disease" | "infection";
  condition_name: string;
  source?: string | null;
  duration?: string | null;
  summary: string;
  damage?: number;
  max_hp_reduction?: number;
  minutes?: number;
  visibility?: ResolutionVisibility;
};
export const previewAffliction = (campaignId: string, input: AfflictionInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/afflictions/preview`, { method: "POST", body: input });
export const confirmAffliction = (campaignId: string, input: Confirmed<AfflictionInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/afflictions/confirm`, { method: "POST", body: input });

export type DowntimeResolutionInput = {
  activity_version: number;
  character_version: number;
  progress_days?: number;
  xp_award?: number;
  summary: string;
  visibility?: ResolutionVisibility;
};
export const previewDowntimeResolution = (campaignId: string, activityId: string, input: DowntimeResolutionInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/downtime/${activityId}/preview`, { method: "POST", body: input });
export const confirmDowntimeResolution = (campaignId: string, activityId: string, input: Confirmed<DowntimeResolutionInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/downtime/${activityId}/confirm`, { method: "POST", body: input });

export type SocialInteractionInput = {
  npc_version: number;
  outcome: "improve" | "unchanged" | "worsen";
  minutes?: number;
  summary: string;
  memory_kind?: "conversation" | "bargain" | "deception" | "intimidation" | "favor" | "other";
  tags?: string[];
  secret?: boolean;
};
export const previewSocialInteraction = (campaignId: string, npcId: string, input: SocialInteractionInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/npcs/${npcId}/social/preview`, { method: "POST", body: input });
export const confirmSocialInteraction = (campaignId: string, npcId: string, input: Confirmed<SocialInteractionInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/npcs/${npcId}/social/confirm`, { method: "POST", body: input });

export type NpcMoraleInput = {
  npc_version: number;
  outcome: "hold" | "retreat" | "surrender";
  combat_id?: string | null;
  combat_version?: number | null;
  leave_combat?: boolean;
  minutes?: number;
  summary: string;
  visibility?: ResolutionVisibility;
};
export const previewNpcMorale = (campaignId: string, npcId: string, input: NpcMoraleInput) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/npcs/${npcId}/morale/preview`, { method: "POST", body: input });
export const confirmNpcMorale = (campaignId: string, npcId: string, input: Confirmed<NpcMoraleInput>) =>
  apiFetch<Record<string, unknown>>(`/campaigns/${campaignId}/npcs/${npcId}/morale/confirm`, { method: "POST", body: input });

export async function listSceneParticipants(
  campaignId: string,
  sceneId: string,
  signal?: AbortSignal,
): Promise<SceneParticipant[]> {
  return (
    await apiFetch<Items<SceneParticipant>>(
      `/campaigns/${campaignId}/scenes/${sceneId}/participants`,
      { signal },
    )
  ).items;
}

export function addSceneParticipant(
  campaignId: string,
  sceneId: string,
  input: {
    entity_type: "character" | "npc" | "monster";
    entity_id: string;
    role?: string;
    visible?: boolean;
  },
): Promise<SceneParticipant> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/participants`, {
    method: "POST",
    body: input,
  });
}

export function removeSceneParticipant(
  campaignId: string,
  sceneId: string,
  participantId: string,
  version: number,
): Promise<void> {
  return apiFetch(
    `/campaigns/${campaignId}/scenes/${sceneId}/participants/${participantId}?version=${version}`,
    { method: "DELETE" },
  );
}

export function startSceneCombat(
  campaignId: string,
  sceneId: string,
  name?: string,
): Promise<SceneCombatResult> {
  return apiFetch(`/campaigns/${campaignId}/scenes/${sceneId}/start-combat`, {
    method: "POST",
    body: { name: name || null },
  });
}

File: backend/src/dnd_dm_assistant/api/routes/world.py
from __future__ import annotations

from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from sqlalchemy.exc import IntegrityError

from dnd_dm_assistant.api.dependencies import (
    get_campaign_service,
    get_exploration_service,
    get_site_service,
    get_world_generation_service,
    get_world_service,
)
from dnd_dm_assistant.api.schemas import (
    ExplorationConfirmRequest,
    ExplorationPreviewRequest,
    ItemPickupRequest,
    LocationGenerationConfirmRequest,
    LocationGenerationRequest,
    MonsterCreate,
    NPCGenerationRequest,
    SceneCombatStartRequest,
    SceneCreate,
    SceneGridCreate,
    SceneObjectCreate,
    SceneParticipantCreate,
    SceneTokenCreate,
    SiteGenerationConfirmRequest,
    SiteGenerationRequest,
    SiteRoomVisibilityRequest,
    TravelConfirmRequest,
    TravelPreviewRequest,
    WorldItemCreate,
)
from dnd_dm_assistant.application.campaigns import CampaignService, NotFoundError
from dnd_dm_assistant.application.rag import RuntimeUnavailableError
from dnd_dm_assistant.application.world_generation import WorldGenerationService
from dnd_dm_assistant.domain.campaign_state import StateNotFoundError, VersionConflict
from dnd_dm_assistant.domain.world import LocationGenerationPreview, NPCGenerationPreview
from dnd_dm_assistant.infrastructure.database.exploration_service import ExplorationService
from dnd_dm_assistant.infrastructure.database.site_service import SiteService
from dnd_dm_assistant.infrastructure.database.world_service import WorldService

router = APIRouter(prefix="/campaigns/{campaign_id}", tags=["world"])


def _request_id(request: Request) -> str:
    return str(getattr(request.state, "request_id", "unknown"))


def _safe_call(fn: Any) -> Any:
    try:
        return fn()
    except (NotFoundError, StateNotFoundError) as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except VersionConflict as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except IntegrityError as exc:
        raise HTTPException(status_code=422, detail="World constraints were violated") from exc


@router.post("/generate/npc", response_model=NPCGenerationPreview)
async def generate_npc(
    campaign_id: str,
    body: NPCGenerationRequest,
    campaign_service: Annotated[CampaignService, Depends(get_campaign_service)],
    generator: Annotated[WorldGenerationService, Depends(get_world_generation_service)],
) -> NPCGenerationPreview:
    campaign = _safe_call(lambda: campaign_service.get("campaign", campaign_id))
    try:
        return await generator.generate_npc(
            campaign=campaign,
            mode=body.mode,
            brief=body.brief,
            answers=body.answers,
        )
    except RuntimeUnavailableError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/generate/location", response_model=LocationGenerationPreview)
async def generate_location(
    campaign_id: str,
    body: LocationGenerationRequest,
    campaign_service: Annotated[CampaignService, Depends(get_campaign_service)],
    generator: Annotated[WorldGenerationService, Depends(get_world_generation_service)],
) -> LocationGenerationPreview:
    campaign = _safe_call(lambda: campaign_service.get("campaign", campaign_id))
    try:
        return await generator.generate_location(
            campaign=campaign,
            brief=body.brief,
            maximum_depth=body.maximum_depth,
            scale=body.scale,
        )
    except RuntimeUnavailableError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/generate/location/confirm", status_code=status.HTTP_201_CREATED)
def confirm_location_generation(
    campaign_id: str,
    body: LocationGenerationConfirmRequest,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.confirm_location_tree(
            campaign_id,
            body.preview.root,
            maximum_depth=body.preview.maximum_depth,
            request_id=_request_id(request),
        )
    )


@router.post("/sites/generate/preview")
def preview_site_generation(
    campaign_id: str,
    body: SiteGenerationRequest,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.preview(campaign_id, body.model_dump()))


@router.post("/sites/generate/confirm", status_code=status.HTTP_201_CREATED)
def confirm_site_generation(
    campaign_id: str,
    body: SiteGenerationConfirmRequest,
    request: Request,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.confirm(campaign_id, body.preview, request_id=_request_id(request))
    )


@router.get("/region-maps")
def list_region_maps(
    campaign_id: str,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return {"region_maps": _safe_call(lambda: service.list_region_maps(campaign_id))}


@router.get("/sites")
def list_adventure_sites(
    campaign_id: str,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return {"sites": _safe_call(lambda: service.list_sites(campaign_id))}


@router.post("/sites/repair-scene-atoms")
def repair_site_scene_atoms(
    campaign_id: str,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, int]:
    return _safe_call(lambda: service.repair_generated_scene_atoms(campaign_id))


@router.get("/sites/{site_id}")
def get_adventure_site(
    campaign_id: str,
    site_id: str,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.get(campaign_id, site_id))


@router.put("/sites/{site_id}/levels/{level_index}/rooms/{room_index}/visibility")
def set_site_room_visibility(
    campaign_id: str,
    site_id: str,
    level_index: int,
    room_index: int,
    body: SiteRoomVisibilityRequest,
    request: Request,
    service: Annotated[SiteService, Depends(get_site_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.set_room_visibility(
            campaign_id,
            site_id,
            level_index,
            room_index,
            visible=body.visible,
            request_id=_request_id(request),
        )
    )


@router.delete("/sites/{site_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_adventure_site(
    campaign_id: str,
    site_id: str,
    request: Request,
    response: Response,
    service: Annotated[SiteService, Depends(get_site_service)],
    version: int = Query(ge=1),
) -> None:
    _safe_call(
        lambda: service.delete(
            campaign_id,
            site_id,
            expected_version=version,
            request_id=_request_id(request),
        )
    )
    response.status_code = status.HTTP_204_NO_CONTENT


@router.get("/items")
def list_items(
    campaign_id: str,
    service: Annotated[WorldService, Depends(get_world_service)],
    location_id: str | None = None,
    owner_character_id: str | None = None,
) -> dict[str, Any]:
    return {
        "items": _safe_call(
            lambda: service.list_items(
                campaign_id,
                location_id=location_id,
                owner_character_id=owner_character_id,
            )
        )
    }


@router.post("/items", status_code=status.HTTP_201_CREATED)
def create_item(
    campaign_id: str,
    body: WorldItemCreate,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.create_item(
            campaign_id,
            body.model_dump(exclude_unset=True),
            request_id=_request_id(request),
        )
    )


@router.delete("/items/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_item(
    campaign_id: str,
    item_id: str,
    request: Request,
    response: Response,
    service: Annotated[WorldService, Depends(get_world_service)],
    version: int = Query(ge=1),
) -> None:
    _safe_call(
        lambda: service.delete_item(
            campaign_id,
            item_id,
            expected_version=version,
            request_id=_request_id(request),
        )
    )
    response.status_code = status.HTTP_204_NO_CONTENT


@router.post("/items/{item_id}/pickup")
def pickup_item(
    campaign_id: str,
    item_id: str,
    body: ItemPickupRequest,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    moved, inventory = _safe_call(
        lambda: service.pickup_item(
            campaign_id,
            item_id,
            character_id=body.character_id,
            quantity=body.quantity,
            expected_version=body.version,
            request_id=_request_id(request),
        )
    )
    return {"item": moved, "inventory": inventory}


@router.get("/characters/{character_id}/inventory")
def get_inventory(
    campaign_id: str,
    character_id: str,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.inventory(campaign_id, character_id))


@router.get("/monsters")
def list_monsters(
    campaign_id: str,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return {"items": _safe_call(lambda: service.list_monsters(campaign_id))}


@router.post("/monsters", status_code=status.HTTP_201_CREATED)
def create_monster(
    campaign_id: str,
    body: MonsterCreate,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.create_monster(
            campaign_id,
            body.model_dump(exclude_unset=True),
            request_id=_request_id(request),
        )
    )


@router.get("/scenes")
def list_scenes(
    campaign_id: str,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return {"items": _safe_call(lambda: service.list_scenes(campaign_id))}


@router.post("/scenes", status_code=status.HTTP_201_CREATED)
def create_scene(
    campaign_id: str,
    body: SceneCreate,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.create_scene(
            campaign_id,
            body.model_dump(exclude_unset=True),
            request_id=_request_id(request),
        )
    )


@router.get("/scenes/{scene_id}/grid")
def get_scene_grid(
    campaign_id: str,
    scene_id: str,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.grid(campaign_id, scene_id))


@router.post("/scenes/{scene_id}/grid", status_code=status.HTTP_201_CREATED)
def create_scene_grid(
    campaign_id: str,
    scene_id: str,
    body: SceneGridCreate,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.create_grid(campaign_id, scene_id, body.model_dump()))


@router.post("/scenes/{scene_id}/tokens", status_code=status.HTTP_201_CREATED)
def create_scene_token(
    campaign_id: str,
    scene_id: str,
    body: SceneTokenCreate,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.add_token(campaign_id, scene_id, body.model_dump()))


@router.post("/scenes/{scene_id}/objects", status_code=status.HTTP_201_CREATED)
def create_scene_object(
    campaign_id: str,
    scene_id: str,
    body: SceneObjectCreate,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.add_object(campaign_id, scene_id, body.model_dump()))


@router.post("/scenes/{scene_id}/exploration/preview")
def preview_exploration(
    campaign_id: str,
    scene_id: str,
    body: ExplorationPreviewRequest,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.preview_exploration(campaign_id, scene_id, body.model_dump()))


@router.post("/scenes/{scene_id}/exploration/confirm")
def confirm_exploration(
    campaign_id: str,
    scene_id: str,
    body: ExplorationConfirmRequest,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.confirm_exploration(campaign_id, scene_id, body.model_dump()))


@router.post("/travel/preview")
def preview_travel(
    campaign_id: str,
    body: TravelPreviewRequest,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.preview_travel(campaign_id, body.model_dump()))


@router.post("/travel/confirm")
def confirm_travel(
    campaign_id: str,
    body: TravelConfirmRequest,
    service: Annotated[ExplorationService, Depends(get_exploration_service)],
) -> dict[str, Any]:
    return _safe_call(lambda: service.confirm_travel(campaign_id, body.model_dump()))


@router.get("/scenes/{scene_id}/participants")
def list_scene_participants(
    campaign_id: str,
    scene_id: str,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return {"items": _safe_call(lambda: service.list_participants(campaign_id, scene_id))}


@router.post("/scenes/{scene_id}/participants", status_code=status.HTTP_201_CREATED)
def add_scene_participant(
    campaign_id: str,
    scene_id: str,
    body: SceneParticipantCreate,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.add_participant(
            campaign_id,
            scene_id,
            body.model_dump(exclude_unset=True),
            request_id=_request_id(request),
        )
    )


@router.delete(
    "/scenes/{scene_id}/participants/{participant_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def remove_scene_participant(
    campaign_id: str,
    scene_id: str,
    participant_id: str,
    request: Request,
    response: Response,
    service: Annotated[WorldService, Depends(get_world_service)],
    version: int = Query(ge=1),
) -> None:
    _safe_call(
        lambda: service.remove_participant(
            campaign_id,
            scene_id,
            participant_id,
            expected_version=version,
            request_id=_request_id(request),
        )
    )
    response.status_code = status.HTTP_204_NO_CONTENT


@router.post("/scenes/{scene_id}/start-combat", status_code=status.HTTP_201_CREATED)
def start_scene_combat(
    campaign_id: str,
    scene_id: str,
    body: SceneCombatStartRequest,
    request: Request,
    service: Annotated[WorldService, Depends(get_world_service)],
) -> dict[str, Any]:
    return _safe_call(
        lambda: service.start_combat(
            campaign_id,
            scene_id,
            name=body.name,
            request_id=_request_id(request),
        )
    )
